applauncher/004-three-by-three.qml

```diff
@@ -0,0 +1,358 @@
+            launcherOuterColor = alb.outerColor(launcherModel.get(0).filePath);
+
+            toLeftAllowed = false
+            toRightAllowed = false
+            toBottomAllowed =  Qt.binding(function() { return !atYBeginning })
+            toTopAllowed = Qt.binding(function() { return !atYEnd })
+            forbidTop = Qt.binding(function() { return !atYBeginning })
+            forbidBottom = false
+            forbidLeft = false
+            forbidRight = false
+        }
+
+        onContentYChanged: {
+            var lowerStop = Math.floor(contentY/appsView.height)
+            var upperStop = lowerStop+1
+            var ratio = (contentY%appsView.height)/appsView.height
+
+            if(upperStop + 1 > launcherModel.itemCount || ratio == 0) {
+                launcherCenterColor = alb.centerColor(launcherModel.get(lowerStop).filePath);
+                launcherOuterColor = alb.outerColor(launcherModel.get(lowerStop).filePath);
+                return;
+            }
+
+            if(lowerStop < 0) {
+                launcherCenterColor = alb.centerColor(launcherModel.get(0).filePath);
+                launcherOuterColor = alb.outerColor(launcherModel.get(0).filePath);
+                return;
+            }
+
+            var upperCenterColor = alb.centerColor(launcherModel.get(upperStop).filePath);
+            var lowerCenterColor = alb.centerColor(launcherModel.get(lowerStop).filePath);
+
+            launcherCenterColor = Qt.rgba(
+                        upperCenterColor.r * ratio + lowerCenterColor.r * (1-ratio),
+                        upperCenterColor.g * ratio + lowerCenterColor.g * (1-ratio),
+                        upperCenterColor.b * ratio + lowerCenterColor.b * (1-ratio)
+                    );
+
+            var upperOuterColor = alb.outerColor(launcherModel.get(upperStop).filePath);
+            var lowerOuterColor = alb.outerColor(launcherModel.get(lowerStop).filePath);
+
+            launcherOuterColor = Qt.rgba(
+                        upperOuterColor.r * ratio + lowerOuterColor.r * (1-ratio),
+                        upperOuterColor.g * ratio + lowerOuterColor.g * (1-ratio),
+                        upperOuterColor.b * ratio + lowerOuterColor.b * (1-ratio)
+                    );
+
+            currentPos = Math.round(lowerStop+ratio)
+        }
+    }
+
+    MouseArea {
+        id: absoluteMouse
+
+        anchors.fill: parent
+        propagateComposedEvents: true
+
+        onPressed: {
+            clickY = Math.round(mouse.y)
+            mouse.accepted = false
+        }
+        onClicked: mouse.accepted = false
+        onReleased: mouse.accepted = false
+        onDoubleClicked: mouse.accepted = false
+        onPositionChanged: mouse.accepted = false
+        onPressAndHold: mouse.accepted = false
+    }
+
+    Text {
+        id: hoverTitle
+
+        property string tempTitle: appTitle
+        property real hoverTextOffset: root.clickY > Dims.h(48) ? -Dims.h(62) : Dims.h(62)
```

Comment:
Avoiding complex position calculation is best. Can’t we have a fixed height and anchors either the top or the bottom to the top or bottom of the parent ? I feel this would be a lot more readable

Reply:
Due to the items being animated, it is much more complex to re-anchor them on top and bottom then the give a negative or positive offset. the code i came up with was much less readable at least and i'd like to keep using the offset. I have added comments to hopefully make the syntax easier to read.